docs/source/high-level-graphs.rst

```diff
@@ -0,0 +1,193 @@
+thousands of tasks per layer) and do perform high-level optimizations.  For
+example in the case above we may want to automatically rewrite our code to
+filter our datasets before adding 100
+
+.. code-block:: python
+
+    # Before
+    df = dd.read_csv('myfile.*.csv')
+    df = df + 100
+    df = df[df.name == 'Alice']
+
+    # After
+    df = dd.read_csv('myfile.*.csv')
+    df = df[df.name == 'Alice']
+    df = df + 100
+
+Dask's high level graphs help us to explicitly encode this structure by storing
+our task graphs in layers, with dependencies between layers.
+
+.. code-block:: python
+
+   >>> import dask.dataframe as dd
+
+   >>> df = dd.read_csv('myfile.*.csv')
+   >>> df = df + 100
+   >>> df = df[df.name == 'Alice']
+
+   >>> graph = df.__dask_graph__()
+   >>> graph.layers
+   {
+    'read-csv': {('read-csv', 0): (pandas.read_csv, 'myfile.0.csv'),
+                 ('read-csv', 1): (pandas.read_csv, 'myfile.1.csv'),
+                 ('read-csv', 2): (pandas.read_csv, 'myfile.2.csv'),
+                 ('read-csv', 3): (pandas.read_csv, 'myfile.3.csv')},
+
+    'add': {('add', 0): (operator.add, 'myfile.0.csv', 100),
+            ('add', 1): (operator.add, 'myfile.1.csv', 100),
+            ('add', 2): (operator.add, 'myfile.2.csv', 100),
+            ('add', 3): (operator.add, 'myfile.3.csv', 100)}
+
+    'filter': {('filter', 0): (lambda part: part[part.name == 'Alice'], ('add', 0)),
+               ('filter', 1): (lambda part: part[part.name == 'Alice'], ('add', 1)),
+               ('filter', 2): (lambda part: part[part.name == 'Alice'], ('add', 2)),
+               ('filter', 3): (lambda part: part[part.name == 'Alice'], ('add', 3))}
+   }
+
+   >>> graph.dependencies
+   {
+    'read-csv': set(),
+    'add': {'read-csv'},
+    'filter': {'add'}
+   }
+
+While the dataframe points to the output layers on which it depends directly
+
+.. code-block:: python
+
+   >>> df.__dask_layers__()
+   {'filter'}
+
+
+HighLevelGraphs
+---------------
+
+The :obj:`HighLevelGraph` object is a ``Mapping`` object composed of other
+sub-``Mappings``, along with a high-level dependency mapping between them.
+
+.. code-block:: python
+
+   class HighLevelGraph(Mapping):
+       layers: Dict[str, Mapping]
+       dependencies: Dict[str, Set[str]]
+
+You can construct a HighLevelGraph explicilty by providing both to the
+constructor.
+
+.. code-block:: python
+
+   layers = {
+      'read-csv': {('read-csv', 0): (pandas.read_csv, 'myfile.0.csv'),
+                   ('read-csv', 1): (pandas.read_csv, 'myfile.1.csv'),
+                   ('read-csv', 2): (pandas.read_csv, 'myfile.2.csv'),
+                   ('read-csv', 3): (pandas.read_csv, 'myfile.3.csv')},
+
+      'add': {('add', 0): (operator.add, 'myfile.0.csv', 100),
+              ('add', 1): (operator.add, 'myfile.1.csv', 100),
+              ('add', 2): (operator.add, 'myfile.2.csv', 100),
+              ('add', 3): (operator.add, 'myfile.3.csv', 100)}
```

Comment:
As in the `HighLevelGraph` docstring, guessing `'myfile.*.csv'` should be `('read-csv', *)`.